[SEC-179] Add annotations to java_class_typet

NathanJPhillips opened the pull request:
Add the annotations parsed from the byte code to classes/fields/methods and static fields (globals) in the symbol table
This could be generally useful, especially in the short term for @allredj's team.

NathanJPhillips committed:
Add annotations to java_class_typet, its methods and fields

Add the annotations parsed from the byte code to classes/fields/methods and static fields (globals) in the symbol table

NathanJPhillips committed:
Add test for added annotations

tautschnig commented on src/java_bytecode/java_types.h:
```diff
@@ -18,6 +18,47 @@ Author: Daniel Kroening, kroening@kroening.com
+    valuet(irep_idt name, exprt value)
+    {
+      id(name);
+      set(ID_value, value);
```
I don't know whether you expect these annotations to grow; if you don't, then I'd be more conservative on memory and do `class valuet : public exprt` with `valuet(const irep_idt &name, const exprt &value) : exprt(name) {  copy_to_operands(value); }`, with `const exprt &get_value() { return op0(); }`.

smowton replied:
Ah yes, agreed, make the value an operand rather than named-operand (vector vs. map allocation).

tautschnig commented on src/java_bytecode/java_types.h:
```diff
@@ -18,6 +18,47 @@ Author: Daniel Kroening, kroening@kroening.com
+    {
+      return static_cast<const exprt &>(find(ID_value));
+    }
+  };
```
I note that `java_annotationt` has no `id()` set?!

smowton replied:
That's in common with many ireps whose ID is implied by its position in the irep graph though, for example struct components.

tautschnig commented on src/java_bytecode/java_types.h:
```diff
@@ -18,6 +18,47 @@ Author: Daniel Kroening, kroening@kroening.com
+  explicit java_annotationt(const typet &type)
+  {
+    set(ID_type, type);
+    add(ID_values);
```
Equally, I'd do `class java_annotationt : public exprt` with `get_values()` just doing `return operands();`.

tautschnig reviewed (commented):
You may consider avoiding maps-of-maps if all you have is a single entry. But just thoughts, you might have very different plans for the data layout.

smowton reviewed (approved):
Agree with @tautschnig's compaction plan, otherwise lgtm

NathanJPhillips commented:
I've not used `exprt` as a base since `java_annotationt` doesn't have operands and `java_annotationt::valuet` doesn't have a type. Essentially neither of them are actually expressions. But I've gained the equivalent efficiency by implementing the equivalent operations by hand.

tautschnig commented on src/java_bytecode/java_types.h:
```diff
@@ -44,6 +91,28 @@ inline java_class_typet &to_java_class_type(typet &type)
+  std::vector<java_annotationt> &get_annotations()
+  {
+    // This cast should be safe as java_annotationt doesn't add data to irept
+    return reinterpret_cast<std::vector<java_annotationt> &>(
```
+1 for using `reinterpret_cast` (and thus something that one can efficiently search for) instead of the C-style cast. Really, we should rid of all of those...

LAJW replied:
This makes me shiver. I'd much prefer a loop/`std::transform` since this bit shouldn't affect performance (in practice it would be iterated over only once). But at least it's not a c-style cast.

tautschnig replied:
Just curious: how would that work as really you want to return a reference to the same storage?

LAJW commented on src/java_bytecode/java_types.h:
```diff
@@ -18,6 +18,46 @@ Author: Daniel Kroening, kroening@kroening.com
+      get_sub().push_back(value);
+    }
+
+    const irep_idt &get_name()
```
Could this method be `const`?

LAJW commented on src/java_bytecode/java_types.h:
```diff
@@ -18,6 +18,46 @@ Author: Daniel Kroening, kroening@kroening.com
+      return id();
+    }
+
+    const exprt &get_value()
```
Could this method be `const`?

LAJW commented on src/java_bytecode/java_types.h:
```diff
@@ -18,6 +18,46 @@ Author: Daniel Kroening, kroening@kroening.com
+
+    const exprt &get_value()
+    {
+      return static_cast<const exprt &>(get_sub()[0]);
```
I'm not a fan of magic constants. That includes magic index access.

NathanJPhillips replied:
Please suggest alternative code.

tautschnig replied:
@LAJW Would `.front()` be more adequate?

LAJW replied:
Nope. I'd like an explanation of what the first sub is. Add documentation at the top of the class specifying which ID does what; is there only one; etc. (there's none ATM).

But this is the problem with `irept`-based classes in general. There are no member variable names, only indices. There's no documentation for indices. In regular classes, even if there's no documentation, you can look at the header and see all members and their types at a glance. Here, the best we can do is just add documentation.

Also, an idea. Use constants named after the intended type: `static constexpr size_t FIELD_NAME=0, ...`. That way at least we'd get variable names.

NathanJPhillips replied:
He wants this:
```
      get_sub() = irept::subt(value_id + 1);
      get_sub()[value_id] = value;
```
Unfortunately this is not a good idea. More unfortunately it's proving difficult to convince him of this.

LAJW replied:
In this warped world of `irept` it is a goo idea.

LAJW replied:
Example:

```c++
enum {
  key_id,
  value_id,
  member_count
}

subs = vector<...>(member_count);
subs[key_id] = key;
subs[value_id] = value;

get_key() {
  return subs[key_id];
}

get_value() {
  return subs[value_id];
}
```

That way it would be almost like:

```c++
struct {
 string key;
 string value;
 getKey() { return key; }
 getValue() { return value; }
};
```

What we have now is:

```c++
struct {
  string v1;
  string v2;
  getKey() { return v1; }
  getValue() { return v2; }
}
```

This pattern is scalable to any number of members. Using .front()/push_back() is not and obstructs the structure even more.

LAJW replied:
Reminds me of this: https://github.com/raxod502/TerrariaClone

> Frankly horrifying inline data tables:
> `static boolean[] solid = {false, true, true, true, true, true, true, ...`

LAJW replied:
But I don't care all that much, since I'm not the author, hence my approval.

tautschnig replied:
It might be redundant, but back to basics: there are multiple options for implementing expression trees. One option is to have a class for each kind of expression, with methods and fields specific to each kind of class. Another option is to have one very generic underlying storage and just put an API on top of that for each kind of expression. The latter provides less insight to the compiler (and thus less type checking), but is easier to extend. The former requires dedicated serialisation, deserialisation, hashing, etc. With `irept` we've got the latter. The ESBMC folks have by now moved over from `irept` to dedicated, fixed classes. You might check with @lucasccordeiro on the advantages and disadvantages of doing so.

For as long as we stick with `irept`, the purpose of a class such as the above is to provide a clear API and hide all implementation details (such as the placement of members in the internal storage). While and index `0` might be magic, it is also an information that is very local. The full scope of where one needs to know this fits on a single screen. Have we always got it perfectly right? No, there have been bugs (I do recall c33266a6f, maybe there have been others). The idea of using an enum instead of a magic number is probably a good one, as long as it doesn't get in the way of keeping the code compact.

LAJW commented on src/java_bytecode/java_types.h:
```diff
@@ -18,6 +18,46 @@ Author: Daniel Kroening, kroening@kroening.com
+    set(ID_type, type);
+  }
+
+  const typet &get_type()
```
Could this method be `const`?

LAJW commented on src/java_bytecode/java_bytecode_convert_class.cpp:
```diff
@@ -729,6 +747,32 @@ static void find_and_replace_parameters(
+  const java_bytecode_parse_treet::annotationst &parsed_annotations,
+  std::vector<java_annotationt> &annotations)
+{
+  for(const java_bytecode_convert_classt::annotationt &annotation
```
I'd use `auto` as the type name is exceptionally verbose (unless the interface is different).

LAJW commented on src/java_bytecode/java_bytecode_convert_class.cpp:
```diff
@@ -729,6 +747,32 @@ static void find_and_replace_parameters(
+      annotation.element_value_pairs.end(),
+      std::back_inserter(values),
+      [](
+        const java_bytecode_convert_classt::annotationt::element_value_pairt
```
You could use `const decltype(annotation.element_value_pairs)::value_type &`.

LAJW commented on src/java_bytecode/java_bytecode_convert_class.cpp:
```diff
@@ -411,8 +424,9 @@ void java_bytecode_convert_classt::convert(
-    class_type.components().push_back(class_typet::componentt());
-    class_typet::componentt &component=class_type.components().back();
+    class_type.components().emplace_back();
+    java_componentt &component =
```
Use `auto`, since the type is already declared on the right-hand side.

LAJW reviewed (approved):
Some minor comments. I'm not a fan of `reinterpret_cast` as performance shouldn't be an issue here. Other than that, it's okay.

tautschnig commented:
I have restarted the Travis job; assuming this was a transient failure, could someone with sufficient insight please make a call whether this needs approval from test-gen or can be merged straightaway?

NathanJPhillips commented:
Created test-gen bump https://github.com/diffblue/test-gen/pull/1522

thk123 reviewed (approved):
Passing TG

NathanJPhillips committed:
Add type_dynamic_cast and friends for java_class_typet

Add type_dynamic_cast and friends for struct_typet, class_typet and java_class_typet

NathanJPhillips committed:
Pretty printing of java_class_typet

Added java_qualifierst that adds annotations as qualifiers to types

thk123 commented:
@NathanJPhillips looks like a real compile error here:
```
/home/travis/build/diffblue/cbmc/src/java_bytecode/java_qualifiers.h:12: undefined reference to `vtable for java_qualifierst'
```

NathanJPhillips commented:
@thk123 indeed - forgot you still need to update Makefile with new files on CBMC.

NathanJPhillips commented:
Could I get at least one re-review on this one? It changed enough while adding tests that I feel it needs it. In particular the commit "Pretty printing of java_class_typet"

smowton commented on src/util/journalling_symbol_table.h:
```diff
@@ -1,4 +1,3 @@
-
```
Probably accidental?

smowton reviewed (approved):
Looks good, if surprisingly involved!

peterschrammel commented on src/java_bytecode/java_qualifiers.h:
```diff
@@ -0,0 +1,38 @@
+
+  java_qualifierst &operator=(const java_qualifierst &other);
+  java_qualifierst &operator=(java_qualifierst &&other);
+  virtual c_qualifierst &operator=(const c_qualifierst &other) override;
```
`virtual` redundant with `override`

NathanJPhillips replied:
This seems like good redundancy from a documentation perspective. Being able to see whether a function is virtual in a single place without having to check the beginning and the end of the declaration is a benefit in my view. As long as it's consistent within a class I think there are examples of both using and not using `virtual` with `override` in CBMC. Do you want to make a general policy on this or accept that it's a style issue on which developers can disagree?

tautschnig replied:
If there is a decision to avoid `virtual ... override` then a7c24fb63 should be revisited: `cpplint.py` would happily warn about those, but this had been commented out (by @peterschrammel  :-) ). Actually there is more to revisit when it comes to a7c24fb63, such as operator spacing...

tautschnig commented on src/ansi-c/c_qualifiers.cpp:
```diff
@@ -9,6 +9,27 @@ Author: Daniel Kroening, kroening@kroening.com
 #include <ostream>
+#include <util/make_unique.h>
+
+c_qualifierst &c_qualifierst::operator=(const c_qualifierst &other)
```
Why would you even want to do this?

NathanJPhillips replied:
To implement `clone` in derived classes we create an instance of the derived object, use this assignment to copy the base class data members to the new instance and then copy the derived class data.

tautschnig commented on src/ansi-c/c_qualifiers.cpp:
```diff
@@ -9,6 +9,27 @@ Author: Daniel Kroening, kroening@kroening.com
+  return *this;
+}
+
+std::unique_ptr<c_qualifierst> c_qualifierst::clone() const
```
What is this good for?

NathanJPhillips replied:
Answered on your comment at the call site.

tautschnig commented on src/ansi-c/c_qualifiers.h:
```diff
@@ -28,7 +29,13 @@ class c_qualifierst
   }
 
-  void clear()
+  c_qualifierst(const c_qualifierst &other) = delete;
```
Why?

NathanJPhillips replied:
To prevent constructing a `c_qualifiert` from a `java_qualifiert` and consequently losing data.

tautschnig commented on src/ansi-c/expr2c.cpp:
```diff
@@ -165,7 +165,8 @@ std::string expr2ct::convert_rec(
 {
-  c_qualifierst new_qualifiers(qualifiers);
+  std::unique_ptr<c_qualifierst> clone = qualifiers.clone();
+  c_qualifierst &new_qualifiers = *clone;
```
?

NathanJPhillips replied:
Not entirely sure what you're asking, so I'll give a full answer, much of which you probably already know.
This code takes the existing qualifiers, adds some more to them and uses them on sub-elements. It doesn't update the qualifiers collection passed in. It therefore needs to make a copy of the qualifiers passed in. The qualifiers are passed by reference to a `c_qualifierst` but might be of a derived type, such as `java_qualifierst`. To allow this we need a virtual "copy" function, which is normally called `clone`. `clone` has to return by reference/pointer and so normally returns a `unique_ptr`. The unique pointer defines the scope of the variable to be this function and handles the destruction of the returned object but within this scope and in order to reduce code churn, it is possible to take a reference to the pointed to value and use it in this function.

tautschnig replied:
Thank you very much for the elaboration! Maybe I was aware of most of that, but having it written down helped nonetheless.

I don't think that having `expr2ct` call functions with a C-specific parameter type is ideal for re-use, but it should be possible to resolve most of that through double dispatch. Let me try to make detailed comments to that effect in the individual places.

NathanJPhillips replied:
I agree that the type of the parameter to a virtual function intended to be overridden for various languages should be a non-language-specific qualifiers class and I can add such an abstract base class if you like, deriving `c_qualifierst` from it.

tautschnig replied:
My apologies that I have not actually come back with a double-dispatch solution. I have started working on one, noticed that this was way more work than I had initially thought, and then got dragged into another matter.

Here are my updated thoughts: I concur that there should be a non-language-specific parameter type. Since you can't do a virtual copy constructor, you might anyhow end up with the `clone` approach that you have already implemented. What I do hope to be avoidable is the `dynamic_cast`, which would be avoided by calling a virtual member of the parameter. All that is still a bit cumbersome, so you might actually consider implementation qualifiers via `irept`, which would make it easy to customise and would avoid all the clone/cast games.

NathanJPhillips replied:
I like your thinking - double dispatch is always my preference over `dynamic_cast` - but I'm not sure the `dynamic_cast` can be avoided in this code - a good place to start looking is `java_qualifierst::operator==` - imagine how you'd implement that without `dynamic_cast`. Using `irept` doesn't avoid the problem, it just puts the class identifier into the `irep::id` field rather than RTTI - you'd still need to check it to do the comparison.

tautschnig replied:
Yes, all agreed. Overall it seems pretty painful to me and I am wondering whether you might just want a `type_qualifierst` that isn't exactly language-agnostic, but instead caters for several languages. For a start you might just put all the extra information that you added to `java_qualifierst` into `c_qualifierst` and reduce the size of your patch a fair bit.

NathanJPhillips replied:
I've added https://github.com/diffblue/cbmc/pull/1831/commits/2162f8ae8d546bc35cfcedc0f2e9434b7ac9d5bc, which brings in a base class, though it doesn't do everything we'd like. If you need more than this I'll need some more pointers.

NathanJPhillips replied:
Our comments seem to have crossed. Would you prefer I swap all this out for a single type_qualifiert class now? I can see pros and cons both ways.

tautschnig replied:
I think (but you might convince me otherwise) that a single `type_qualifiert` class would make the code a lot simpler, yes.

NathanJPhillips replied:
Oh no, it would definitely make the code simpler. But it would add Java specific stuff to a class used by CBMC. Pros and cons...
Does @peterschrammel have an opinion?

tautschnig commented on src/cpp/expr2cpp.cpp:
```diff
@@ -132,7 +132,8 @@ std::string expr2cppt::convert_rec(
 {
-  c_qualifierst new_qualifiers(qualifiers);
+  std::unique_ptr<c_qualifierst> clone = qualifiers.clone();
+  c_qualifierst &new_qualifiers = *clone;
```
?

tautschnig commented on src/java_bytecode/expr2java.h:
```diff
@@ -20,30 +20,29 @@ class expr2javat:public expr2ct
+  std::string convert_code_java_delete(const exprt &src, unsigned precedence);
+  virtual std::string convert_struct(
+    const exprt &src, unsigned &precedence) override;
+  virtual std::string convert_code(const codet &src, unsigned indent) override;
```
How do those changes relate to the commit subject?

tautschnig commented on src/java_bytecode/expr2java.h:
```diff
@@ -20,30 +20,29 @@ class expr2javat:public expr2ct
     const typet &src,
     const c_qualifierst &qualifiers,
-    const std::string &declarator);
+    const std::string &declarator) override;
```
Using both `virtual` and `override` is redundant. Applies above as well.

tautschnig commented on src/java_bytecode/java_qualifiers.cpp:
```diff
@@ -0,0 +1,154 @@
+// Copyright 2018 DiffBlue Limited. All Rights Reserved.
```
Not consistent with #1919.

tautschnig commented on src/java_bytecode/java_qualifiers.cpp:
```diff
@@ -0,0 +1,154 @@
+{
+  INVARIANT(
+    &other.ns == &ns,
+    "Can only assign from a java_qualifierst using the same namespace");
```
Reaffirms that you don't want the `operator=`.

tautschnig commented on src/java_bytecode/java_qualifiers.cpp:
```diff
@@ -0,0 +1,154 @@
+{
+  INVARIANT(
+    &other.ns == &ns,
+    "Can only assign from a java_qualifierst using the same namespace");
```
See above.

tautschnig commented on src/java_bytecode/java_qualifiers.cpp:
```diff
@@ -0,0 +1,154 @@
+
+c_qualifierst &java_qualifierst::operator=(const c_qualifierst &c_other)
+{
+  auto other = dynamic_cast<const java_qualifierst *>(&c_other);
```
You really don't want this operator.

tautschnig commented on src/java_bytecode/java_qualifiers.cpp:
```diff
@@ -0,0 +1,154 @@
+  return *this;
+}
+
+std::unique_ptr<c_qualifierst> java_qualifierst::clone() const
```
Why?

tautschnig commented on src/java_bytecode/java_qualifiers.cpp:
```diff
@@ -0,0 +1,154 @@
+void java_qualifierst::read(const typet &src)
+{
+  c_qualifierst::read(src);
+  auto &annotated_type = static_cast<const annotated_typet &>(src);
```
There should be a `to_annotated_type`.

NathanJPhillips replied:
Even if there's no possible run-time check for this?

tautschnig replied:
Yes, because it puts all the type casts in one place where maybe, some day, the type hierarchy can be changed. Or additional checks can be added.

tautschnig commented on src/java_bytecode/java_qualifiers.cpp:
```diff
@@ -0,0 +1,154 @@
+void java_qualifierst::write(typet &src) const
+{
+  c_qualifierst::write(src);
+  auto &annotated_type = static_cast<annotated_typet &>(src);
```
`to_annotated_type`

tautschnig commented on src/java_bytecode/java_qualifiers.cpp:
```diff
@@ -0,0 +1,154 @@
+  annotated_type.get_annotations() = annotations;
+}
+
+c_qualifierst &java_qualifierst::operator+=(const c_qualifierst &c_other)
```
There should be an `operator+=` defined with parameter type `java_qualifierst`. The cast should be unnecessary.

NathanJPhillips replied:
This is an override of the virtual function with this signature on the base class.

tautschnig commented on src/java_bytecode/java_qualifiers.cpp:
```diff
@@ -0,0 +1,154 @@
+  return *this;
+}
+
+bool java_qualifierst::operator==(const c_qualifierst &c_other) const
```
Define `operator==` with parameter type `java_qualifierst`. No need for a dynamic cast.

NathanJPhillips replied:
This is an override of the virtual function with this signature on the base class.
This is needed to prevent a `c_qualifierst` comparing equal to a `java_qualifierst` that has annotations and to consequently to make sure that `operator==` is commutative.

tautschnig commented on src/java_bytecode/java_qualifiers.cpp:
```diff
@@ -0,0 +1,154 @@
+    c_qualifierst::operator==(c_other) && annotations == other->annotations;
+}
+
+bool java_qualifierst::is_subset_of(const c_qualifierst &c_other) const
```
Define it for a parameter type `java_qualifierst`?

NathanJPhillips replied:
This is an override of the virtual function with this signature on the base class.

tautschnig commented on src/java_bytecode/java_qualifiers.h:
```diff
@@ -0,0 +1,38 @@
+// Copyright 2018 DiffBlue Limited. All Rights Reserved.
```
See #1919.

tautschnig reviewed (commented):
I believe the introduction of `java_qualifierst` needs more work. In particular I don't understand the additional assignment and clone operators.

NathanJPhillips committed:
Fix spurious virtual function related keywords

Remove virtual keyword from methods that are specific to this class
Add comment on method that hides a non-virtual base method and remove virtual keyword from it
Mark overrides with override

NathanJPhillips committed:
Add to_annotated_type and enable type_checked_cast for annotated_typet

NathanJPhillips committed:
Introduce abstract qualifierst base class

peterschrammel reviewed (approved):
LGTM, apart from redundant virtual/override
